src/validators/arguments_v3.rs

```diff
@@ -0,0 +1,654 @@
+            "positional_only" => Ok(Self::PositionalOnly),
+            "positional_or_keyword" => Ok(Self::PositionalOrKeyword),
+            "var_args" => Ok(Self::VarArgs),
+            "keyword_only" => Ok(Self::KeywordOnly),
+            "var_kwargs_uniform" => Ok(Self::VarKwargsUniform),
+            "var_kwargs_unpacked_typed_dict" => Ok(Self::VarKwargsUnpackedTypedDict),
+            s => py_schema_err!("Invalid var_kwargs mode: `{}`", s),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Parameter {
+    name: String,
+    mode: ParameterMode,
+    lookup_key_collection: LookupKeyCollection,
+    validator: CombinedValidator,
+}
+
+impl Parameter {
+    fn is_variadic(&self) -> bool {
+        matches!(
+            self.mode,
+            ParameterMode::VarArgs | ParameterMode::VarKwargsUniform | ParameterMode::VarKwargsUnpackedTypedDict
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct ArgumentsV3Validator {
+    parameters: Vec<Parameter>,
+    positional_params_count: usize,
+    loc_by_alias: bool,
+    extra: ExtraBehavior,
+    validate_by_alias: Option<bool>,
+    validate_by_name: Option<bool>,
+}
+
+impl BuildValidator for ArgumentsV3Validator {
+    const EXPECTED_TYPE: &'static str = "arguments-v3";
+
+    fn build(
+        schema: &Bound<'_, PyDict>,
+        config: Option<&Bound<'_, PyDict>>,
+        definitions: &mut DefinitionsBuilder<CombinedValidator>,
+    ) -> PyResult<CombinedValidator> {
+        let py = schema.py();
+
+        let arguments_schema: Bound<'_, PyList> = schema.get_as_req(intern!(py, "arguments_schema"))?;
+        let mut parameters: Vec<Parameter> = Vec::with_capacity(arguments_schema.len());
+
+        let mut had_default_arg = false;
+        let mut had_keyword_only = false;
+
+        for arg in arguments_schema.iter() {
+            let arg = arg.downcast::<PyDict>()?;
+
+            let py_name: Bound<PyString> = arg.get_as_req(intern!(py, "name"))?;
+            let name = py_name.to_string();
+            let py_mode = arg.get_as::<Bound<'_, PyString>>(intern!(py, "mode"))?;
+            let py_mode = py_mode
+                .as_ref()
+                .map(|py_str| py_str.to_str())
+                .transpose()?
+                .unwrap_or("positional_or_keyword");
+
+            let mode = ParameterMode::from_str(py_mode)?;
+
+            if mode == ParameterMode::KeywordOnly {
+                had_keyword_only = true;
+            }
+
+            let schema = arg.get_as_req(intern!(py, "schema"))?;
+
+            let validator = match build_validator(&schema, config, definitions) {
+                Ok(v) => v,
+                Err(err) => return py_schema_err!("Parameter '{}':\n  {}", name, err),
+            };
+
+            let has_default = match validator {
+                CombinedValidator::WithDefault(ref v) => {
+                    if v.omit_on_error() {
+                        return py_schema_err!("Parameter '{}': omit_on_error cannot be used with arguments", name);
```

Comment:
```suggestion
                        return py_schema_err!("Parameter '{name}': omit_on_error cannot be used with arguments");
```
?

Reply:
Indeed, looks like the single-argument form of all these macros assume a `&'static str` without formatting. I had a quick look at changing this, but it looks harder than the effort it's worth. Maybe a future thing.